mm2src/coins/utxo/utxo_common.rs

```diff
@@ -2757,130 +2838,113 @@ pub async fn list_mature_unspent_ordered<'a, T: UtxoCommonOps>(
                     Ok(conf) => tx.confirmations = conf,
                     // do not skip the transaction with unknown confirmations,
                     // because the transaction can be matured
-                    Err(e) => log!((e)),
+                    Err(e) => error!("{}", e),
                 }
                 tx
             },
             VerboseTransactionFrom::Rpc(mut tx) => {
                 if tx.height.is_none() {
                     tx.height = unspent.height;
                 }
-                if let Err(e) = coin.cache_transaction_if_possible(&tx).await {
-                    log!((e));
+                if can_tx_be_cached(&tx) {
+                    txs_to_cache.insert(tx_hash, tx.clone());
                 }
                 tx
             },
         };
 
         if coin.is_unspent_mature(&tx_info) {
-            result.push(unspent);
+            result.mature.push(unspent);
+        } else {
+            result.immature.push(unspent);
         }
     }
 
-    Ok((result, recently_spent))
+    cache_transactions_if_possible(coin.as_ref(), &txs_to_cache).await;
+    Ok(result)
 }
 
 pub fn is_unspent_mature(mature_confirmations: u32, output: &RpcTransaction) -> bool {
     // don't skip outputs with confirmations == 0, because we can spend them
     !output.is_coinbase() || output.confirmations >= mature_confirmations
 }
 
+/// [`UtxoCommonOps::get_verbose_transactions_from_cache_or_rpc`] implementation.
+/// Loads verbose transactions from cache or requests it using RPC client.
 #[cfg(not(target_arch = "wasm32"))]
-pub async fn get_verbose_transaction_from_cache_or_rpc(
+pub async fn get_verbose_transactions_from_cache_or_rpc(
     coin: &UtxoCoinFields,
-    txid: H256Json,
-) -> UtxoRpcResult<VerboseTransactionFrom> {
-    let tx_cache_path = match &coin.tx_cache_directory {
-        Some(p) => p.clone(),
-        _ => {
-            // the coin doesn't support TX local cache, don't try to load from cache and don't cache it
-            let tx = coin.rpc_client.get_verbose_transaction(&txid).compat().await?;
-            return Ok(VerboseTransactionFrom::Rpc(tx));
+    tx_ids: HashSet<H256Json>,
+) -> UtxoRpcResult<HashMap<H256Json, VerboseTransactionFrom>> {
+    let mut result_map = HashMap::with_capacity(tx_ids.len());
+    let mut to_request = Vec::with_capacity(tx_ids.len());
+
+    match coin.tx_cache_directory {
+        Some(ref tx_cache_path) => {
+            tx_cache::load_transactions_from_cache_concurrently(tx_cache_path, tx_ids.into_iter())
+                .await
+                .into_iter()
+                .for_each(|(txid, res)| match res {
```

Comment:
A bit too much nesting here too to my taste.

Reply:
I agree with you! Refactored